docs/content/guides/developer/advanced/randomness-onchain.mdx

```diff
@@ -0,0 +1,140 @@
+...
+	struct GuessedCorrectly has drop { ... };
+
+	/// If you guess correctly the output you get a GuessedCorrectly object.
+	public fun play_dice(guess: u8, fee: Coin<SUI>, r: &Random, ctx: &mut TxContext): Option<GuessedCorrectly> {
+		// Pay for the turn
+		assert!(coin::value(&fee) == 1000000, EInvalidAmount);
+		transfer::public_transfer(fee, CREATOR_ADDRESS);
+		// Roll the dice
+		let generator = new_generator(r, ctx);
+		if (guess == random::generate_u8_in_range(&mut generator, 1, 6)) { 
+			option::some(GuessedCorrectly {})
+		} else {
+			option::none()
+		}
+	}
+...
+}
+```
+
+An attacker can deploy the next function:
+
+```move
+public fun attack(guess: u8, r: &Random, ctx: &mut TxContext): GuessedCorrectly {
+		let output = dice::play_dice(guess, r, ctx);
+		option::extract(output) // reverts the transaction if roll_dice returns option::none()
+}
+```
+
+The attacker can now call `attack` with a guess, and always revert the fee transfer if the guess is incorrect.
+
+To protect against composition attacks in this example, define `play_dice` as a private `entry` function so functions from other modules cannot call it, such as, 
+
+```move
+entry fun play_dice(guess: u8, fee: Coin<SUI>, r: &Random, ctx: &mut TxContext): Option<GuessedCorrectly> {
+		...
+}
+```
+
+:::note
+
+Move compiler enforces this behavior by rejecting `public` functions with `Random` as an argument.
+
+:::
+
+## Programmable transaction block (PTB) restrictions
+
+A similar attack to the one previously described involves PTBs *even* when `play_dice` is defined as a private `entry` function. For example, consider the `entry play_dice(guess: u8, fee: Coin<SUI>, r: &Random, ctx: &mut TxContext): Option<GuessedCorrectly> { … }` function defined earlier, the attacker can publish the function
+
+```move
+public fun attack(output: Option<GuessedCorrectly>): GuessedCorrectly {
+		option::extract(output)
+}
+```
+
+and send a PTB with commands `play_dice(...), attack(Result(0))` where `Result(0)` is the output of the first command. As before, the attack takes advantage of the atomic nature of PTBs and always reverts the *entire transaction* if the guess was incorrect, without paying the fee. Sending multiple transactions can repeat the attck, each one executed with different randomness and reverted if the guess is incorrect.
+
+:::note
+
+To protect against PTB-based composition attacks, Sui rejects PTBs that have commands that are not `TransferObjects` or `MergeCoins` following a `MoveCall` command that uses `Random` as an input.
+
+:::
+
+## Instantiating `RandomGenerator`
+
+`RandomGenerator` is secure as long as it's created by the consuming module. If passed as an argument, the caller might be able to predict the outputs of that `RandomGenerator` instance (for example, by calling `bcs::to_bytes(&generator)` and parsing its internal state).
+
+:::note
+
+The Move compiler prevents defining `public` functions with `RandomGenerator` as an argument.
```

Comment:
maybe let's be consistent with the note from above ("Move compiler enforces this behavior by rejecting public functions with Random as an argument."), either that option or the one here in both